cmd/local-gadget/trace/dns.go

```diff
@@ -0,0 +1,164 @@
+		col.Visible = true
+
+		parser, err := commonutils.NewGadgetParserWithRuntimeInfo(&commonFlags.OutputConfig, cols)
+		if err != nil {
+			return commonutils.WrapInErrParserCreate(err)
+		}
+
+		eventCallback := func(event dnsTypes.Event) {
+			baseEvent := event.GetBaseEvent()
+			if baseEvent.Type != eventtypes.NORMAL {
+				commonutils.ManageSpecialEvent(baseEvent, commonFlags.Verbose)
+				return
+			}
+
+			// TODO: This enrich logic could be moved to the
+			// gadget itself to simplify this here and in
+			// pkg/gadget-collection/gadgets/trace/dns/gadget.go
+			container := localGadgetManager.GetContainer(event.Key)
+			if container != nil {
+				event.Namespace = container.Namespace
+				event.Pod = container.Podname
+				event.Container = container.Name
+			}
+
+			switch commonFlags.OutputMode {
+			case commonutils.OutputModeJSON:
+				b, err := json.Marshal(event)
+				if err != nil {
+					fmt.Fprint(os.Stderr, fmt.Sprint(commonutils.WrapInErrMarshalOutput(err)))
+					return
+				}
+
+				fmt.Println(string(b))
+			case commonutils.OutputModeColumns:
+				fallthrough
+			case commonutils.OutputModeCustomColumns:
+				fmt.Println(parser.TransformIntoColumns(&event))
+			default:
+				fmt.Fprint(os.Stderr, commonutils.WrapInErrOutputModeNotSupported(commonFlags.OutputMode))
+			}
+		}
+
+		tracer, err := dnsTracer.NewTracer()
+		if err != nil {
+			return commonutils.WrapInErrGadgetTracerCreateAndRun(err)
+		}
+		defer tracer.Close()
+
+		if commonFlags.OutputMode != commonutils.OutputModeJSON {
+			fmt.Println(parser.BuildColumnsHeader())
+		}
+
+		netnsHost, _ := containerutils.GetNetNs(os.Getpid())
+		genKey := func(container *containercollection.Container) string {
+			if container.Netns == netnsHost {
+				return "host"
+			}
+			return container.ID
+		}
+
+		attachContainerFunc := func(container *containercollection.Container) {
+			err = tracer.Attach(genKey(container), container.Pid, eventCallback)
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "failed to trace container: %q: %s", container.Name, err)
+			}
+		}
+
+		detachContainerFunc := func(container *containercollection.Container) {
+			err = tracer.Detach(genKey(container))
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "failed to stop tracing container: %q: %s", container.Name, err)
+			}
+		}
+
+		containers := localGadgetManager.ContainerCollection.Subscribe(
+			logalGadgetSubKey,
+			containercollection.ContainerSelector{
+				Name: commonFlags.Containername,
+			},
+			func(event containercollection.PubSubEvent) {
+				switch event.Type {
+				case containercollection.EventTypeAddContainer:
+					attachContainerFunc(event.Container)
+				case containercollection.EventTypeRemoveContainer:
+					detachContainerFunc(event.Container)
+				}
+			},
+		)
+		defer localGadgetManager.ContainerCollection.Unsubscribe(logalGadgetSubKey)
+
+		for _, container := range containers {
+			attachContainerFunc(container)
+			defer detachContainerFunc(container)
```

Comment:
Maybe using `tracer.Detach`? We don't care about printing errors during closure.

```suggestion
			defer tracer.Detach(genKey(container))
```